runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -250,6 +261,228 @@ J9::ValuePropagation::isKnownStringObject(TR::VPConstraint *constraint)
           && (constraint->isConstString() || constraint->getKnownObject());
    }
 
+bool J9::ValuePropagation::transformIndexOfKnownString(
+   TR::Node *indexOfNode,
+   TR::Node *sourceStringNode,
+   TR::Node *targetCharNode,
+   TR::Node *startNode,
+   TR::Node *lengthNode,
+   bool is16Bit)
+   {
+   // Keep track of whether or not all constraints are global.
+   bool isGlobal = true;
+   bool isGlobalQuery;
+
+   TR::VPConstraint *sourceConstraint = getConstraint(sourceStringNode, isGlobalQuery);
+   isGlobal &= isGlobalQuery;
+   if (!sourceConstraint)
+      return false;
+   TR::VPKnownObject *knownObject = sourceConstraint->getKnownObject();
+   // The source string must either be a KnownObject or a ConstString.
+   // Otherwise, do not attempt transformations.
+   if (!knownObject && !sourceConstraint->isConstString())
+      return false;
+   TR::KnownObjectTable *knot;
+   if (knownObject)
+      {
+      knot = comp()->getOrCreateKnownObjectTable();
+      if (!knot)
+         return false;
+      TR_OpaqueClassBlock *klazz = knownObject->getClass();
+      if (!comp()->fej9()->isPrimitiveArray(klazz))
+         return false;
+      }
+
+   TR::VPConstraint *targetConstraint = getConstraint(targetCharNode, isGlobal) ;
+   bool targetIsConstChar = false;
+   uint16_t targetChar = -1;
+   if (!targetConstraint)
+      {
+      targetIsConstChar = false;
+      }
+   else if (targetConstraint->asIntConst())
+      {
+      targetIsConstChar = true;
+      targetChar = targetConstraint->asIntConst()->getInt();
```

Comment:
Can we get here since we are presumably dealing with a char ? If so, maybe targetChar should be declared as 32-bit ?

Reply:
I tried adding some logging and it looks like we are getting there.

For a call like `"abcde".indexOf('a')`, the trees look like this:
```
n5n         icall  java/lang/String.indexOf(I)I[#361  final virtual Method -200] [flags 0x20500 0x0 ]  [0x7f0d05419490] bci=[-1,4,92] rc=2 vc=9 vn=- li=- udi=- nc=2
n3n           aload  <string "abcde">[#358  Static] (obj1) [flags 0x80000307 0x0 ]            [0x7f0d054193f0] bci=[-1,0,92] rc=1 vc=8 vn=- li=- udi=- nc=0
n4n           iconst 97                                                                       [0x7f0d05419440] bci=[-1,2,92] rc=1 vc=8 vn=- li=- udi=- nc=0
```

I think because the `'a'` becomes an `iconst`, the constraint is an integer type.

I will change targetChar to 32-bit.